crates/integrations/datafusion/src/physical_plan/scan.rs

```diff
@@ -0,0 +1,123 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::{any::Any, pin::Pin, sync::Arc};
+
+use datafusion::{
+    arrow::{array::RecordBatch, datatypes::SchemaRef as ArrowSchemaRef},
+    execution::{SendableRecordBatchStream, TaskContext},
+    physical_expr::EquivalenceProperties,
+    physical_plan::{
+        stream::RecordBatchStreamAdapter, DisplayAs, ExecutionMode, ExecutionPlan, Partitioning,
+        PlanProperties,
+    },
+};
+use futures::{Stream, TryStreamExt};
+use iceberg::table::Table;
+
+use crate::to_datafusion_error;
+
+#[derive(Debug)]
+pub(crate) struct IcebergTableScan {
+    table: Table,
+    schema: ArrowSchemaRef,
+    plan_properties: PlanProperties,
+}
+
+impl IcebergTableScan {
+    pub(crate) fn new(table: Table, schema: ArrowSchemaRef) -> Self {
+        let plan_properties = Self::compute_properties(schema.clone());
+
+        Self {
+            table,
+            schema,
+            plan_properties,
+        }
+    }
+
+    fn compute_properties(schema: ArrowSchemaRef) -> PlanProperties {
+        // TODO:
+        // This is more or less a placeholder, to be replaced
+        // once we support output-partitioning
+        PlanProperties::new(
+            EquivalenceProperties::new(schema),
+            Partitioning::UnknownPartitioning(1),
+            ExecutionMode::Bounded,
+        )
+    }
+}
+
+impl ExecutionPlan for IcebergTableScan {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        _children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> datafusion::error::Result<Arc<dyn ExecutionPlan>> {
+        Ok(self)
```

Comment:
I think this should never be called, how about throwing an error here?

Reply:
I thought the same; then I checked `CsvExec` in Datafusion since it's also a leaf-node that should also not have any children. turns out they do the same and return `Ok(self)` as well. [link](https://github.com/apache/datafusion/blob/main/datafusion/core/src/datasource/physical_plan/csv.rs#L181-L186)